-implement the up/down keys for the gallery view/slide modal

dstopp opened the pull request:
- Fixes #715
- Were any Django, import pipeline, table_schema, or dictionary files modified? Y/N
  - If YES:
    - Database used: opus3_test_200903
    - All Django tests pass: NA
    - FLAKE8 run on modified code: NA
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y/NA
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
-Catch the up and down arrows when the modal is open; in gallery view, the up/down will move the binoculars up or down one row and update the modal w/the new opusID; in table view, the up/down will do the same as next/prev and move one up or one down, then update the modal.
-if there are no more rows up or down, the up/down arrow will do nothing.

Known problems:
 - there is a bug in the table view that causes the highlighted row to always be at the bottom of the screen once it reaches a certain part of the visible screen.
 - there is a bug for a corner case when the binoculars or highlighted row are scrolled up off the screen, edit of the metadata fields will cause the metadata on the slide view modal to disappear

dstopp committed:
-implement the up/down keys for the gallery view/slide modal

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -559,23 +536,69 @@ var o_browse = {
                         o_browse.loadPageIfNeeded("prev", detailOpusId);
                         break;
+                    case 38:  // up
+                        // note that if there is no where to go in the up direction, nothing will change ... it could go to the very first, tho
```
I don't understand this comment. What does "it could go to the very first" mean? Also "nowhere" is one word.

dstopp committed:
-prevent default behaviour w/the up/down arrows when the modal is showing so that the modal can take over the up/down

dstopp committed:
-removed unnecessary comments

dstopp committed:
-prevent default on all of the keyboard inputs when the modal is open
-check for valid opusID for the spacebar add to cart - wasn't needed before because we couldn't go that fast, but now it is needed as the up/down arrows traverse the data much faster

dstopp committed:
prevent arrow and space keys while data is loading

dstopp committed:
fixed typo

dstopp committed:
-experimenting w/not letting the user arrow key to the bottom row when a load is occurring

dstopp committed:
remove the extra spinner that was only on the gallery view slide modal and move the general spinner always on top

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,102 @@ var o_browse = {
                 switch (e.which || e.keyCode) {
                     case 32:  // spacebar
-                        if (opus.metadataDetailOpusId !== "") {
+                        if ((opus.metadataDetailOpusId !== "" && opus.metadataDetailOpusId !== undefined) && !$("#galleryViewContents").hasClass("op-disabled")) {
```
No reason to have the first set of nested parens. (A & B) & C is the same as A & B & C.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,102 @@ var o_browse = {
-                        o_browse.loadPageIfNeeded("next", detailOpusId);
+                        obsNum++;
+                        // if this new observation is in the last row of screen, don't allow the user to arrow there
+                        if (!o_browse.isLastRowOfObservations(tab, viewNamespace, obsNum)) {
```
This should only matter when loading is in progress. The way it's written right now, I can never "arrow" to the last row of observations at the end of the result set.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,102 @@ var o_browse = {
+                        break;
+                    case 40:  // down
+                        offset = (o_browse.isGalleryView() ? viewNamespace.galleryBoundingRect.x : 1);
+                        if (obsNum + offset <= viewNamespace.totalObsCount) {
```
Same here - this should only be checked when loading is in progress.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,102 @@ var o_browse = {
         });
     }, // end browse behaviors
 
+    isLastRowOfObservations: function(tab, viewNamespace, obsNum) {
```
I'm confused by this function. The point is not to check for the last row on the current screen. The point is to check for the last row in the cached data regardless of whether it's displayed or not. This appears to be doing the former.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,102 @@ var o_browse = {
+    },
+
+    onGalleryOrRowClick: function(obj, e) {
+        // make sure selected modal thumb is unhighlighted, as clicking on this closes the modal
```
I really hate the use of the word "modal" everywhere. There are dozens of "modals" in OPUS. We should always specify which one we're talking about. Do you mean the "slide show modal"? (I know this was an existing comment that got moved, but still we should fix it)

rfrenchseti reviewed (changes requested):
I would still like the slide show dialog not to recenter when it's open and you click on a new thumbnail. With the full arrows available, moving the dialog out of the way will be more common, and this will show up more.

dstopp committed:
-simplified and looking just at cached observations
-address comments in pull request

dstopp committed:
only recenter the gallery view/slide modal when opening new, not when moving to another observation or clicking on a different observation

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,94 @@ var o_browse = {
-                        o_browse.removeEditMetadataDetails();
-                        o_browse.loadPageIfNeeded("next", detailOpusId);
+                        obsNum++;
+                        // if this new observation is in the last row of cached observations, don't allow the user to arrow there
```
Isn't this comment wrong? I thought we only disallowed the last row if a load was in progress. In fact I don't understand this at all - I can move the binocs to the last row, but this code would seem to prevent that. What am I missing?

rfrenchseti replied:
Also, if you go to the table view, make enough columns that there's a horizontal scrollbar, go to the very last observation, and hit the right arrow, the scrollbar will move rather than just ignoring the arrow like we should because it's now captured by the slide.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,94 @@ var o_browse = {
+                        }
                         break;
                     case 37:  // prev
+                        obsNum--;
```
Why don't we check for obsNum < 1 here and return false like everywhere else in this switch?

dstopp replied:
Even though the check is unnecessary as there are other checks in place for this, i have added a check to the code for consistency.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,94 @@ var o_browse = {
+                            detailOpusId = (o_browse.isGalleryView() ? $(tab).find(`.op-thumbnail-container[data-obs='${obsNum}']`).data("id") : $(tab).find(`tr[data-obs='${obsNum}']`).data("id"));
+                            o_browse.removeEditMetadataDetails();
+                            o_browse.loadPageIfNeeded("prev", detailOpusId);
+                        }
```
Some other cases return false when we don't do anything. Why not here?

dstopp replied:
It's superfluous - there's no change in behaviour if there is a return false here or not, it only matters at the end of the cache where the right arrow or down can start scrolling in some direction that needs to be controlled.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -548,34 +521,94 @@ var o_browse = {
+                        offset = (o_browse.isGalleryView() ? viewNamespace.galleryBoundingRect.x : 1);
+                        if (obsNum + offset <= viewNamespace.totalObsCount) {
+                            obsNum += offset;
+                            // if this new observation is in the last row of cached observations, don't allow the user to arrow there
```
Same comment as above about this comment and only doing on load in progress.

dstopp replied:
comment clarified in both places.

rfrenchseti reviewed (changes requested):
1) There's an existing bug that has gotten worse with these changes:
- Click on a thumbnail
- Enter "Edit" mode on the slide
- Add or remove a field
- While the new metadata is being retrieved, click on a different background thumbnail
- The binoculars will disappear completely (up until now, same behavior as production)
- On the production version, you can use the left/right arrows to get the binocs back
- In the new version, the left arrow works, but right/up/down don't do anything

2) The up/down keys do the same thing (move the binocs) whether the mouse is on top of the slide dialog or the background thumbnails. However, the spacebar (toggle cart) both toggles the cart AND scrolls the browser window if the mouse is over the thumbnails. The default behavior of spacebar needs to be disabled with the slide is up.

dstopp committed:
-reorganize the switch statement that handles the arrow/space key press when the gallery view slide modal is open for consistancy
-fix bug where the binoculars disappear if the modal is open when the infinite scroll page window is fetched
-fix issue where the left/right arrow keys stop working after a new fetch occurs

dstopp committed:
-set the global var that keeps track of the slide view modal to "" when the modal is closed.  This fixes the remnant binoculars added to the screen when new data is fetched <ie., for a change to the metadata>.

juzen2003 reviewed (changes requested):
- Database used: opus3_test_200903
- All Django tests pass: Y
- JSHINT run on all affected files: Y
    - In browse.js
        - undefined variable ```html``` in ```updateGalleryView```
        - unused variable ```obsNum``` at line 2030
- FLAKE8 run on all affected files: Y
- Tested on Chrome, Firefox, Safari, and iOS: Y
  (Remember to test all browser sizes)
- Tested for race conditions (with delayed API calls if applicable): Y

- Only minor changes needed for jshint test on browse.js

dstopp committed:
-minor fixes found in jshint
